src/algebra/commute.lean

```diff
@@ -0,0 +1,250 @@
+/-
+Copyright (c) 2019 Neil Strickland. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Neil Strickland
+
+Commuting pairs of elements in not-necessarily-commutative monoids,
+groups and (semi) rings.  Centralizers as submonoids, subgroups or
+subrings.
+
+Usage example: suppose that a, b and c are elements of a ring, and
+that `hb : commute a b` and `hc : commute a c`.  Then
+`hb.pow_left 5` proves `commute (a ^ 5) b` and
+`(hb.pow 2).add (hb.mul hc)` proves `commute a (b ^ 2 + b * c)`.
+Lean does not immediately recognise these terms as equations,
+so for rewriting we need syntax like `rw [(hb.pow_left 5).eq]`
+rather than just `rw [hb.pow_left 5]`.
+-/
+
+import algebra.group_power
+  group_theory.submonoid group_theory.subgroup
+  ring_theory.subring
+
+section semigroup
+
+variables {S : Type*} [semigroup S]
+
+@[reducible] def commute (a b : S) : Prop := a * b = b * a
+
+theorem commute.eq {a b : S} (h : commute a b) : a * b = b * a := h
+
+@[refl] theorem commute.refl (a : S) : commute a a := by { unfold commute }
+
+@[symm] theorem commute.symm {a b : S} : commute a b → commute b a :=
```

Comment:
You can of course also write
```suggestion
@[symm] theorem commute.symm {a b : S} (h : commute a b) :  commute b a :=
```
I don't care too much...
In the next theorem it might help a bit more.

Reply:
Done.